src/server/worker/datum_iterator.go

```diff
@@ -137,41 +137,44 @@ func (d *unionDatumIterator) Datum() []*Input {
 }
 
 type crossDatumIterator struct {
-	inputs  []DatumIterator
-	started bool
+	iterators []DatumIterator
+	started   bool
 }
 
 func newCrossDatumIterator(pachClient *client.APIClient, cross []*pps.Input) (DatumIterator, error) {
 	result := &crossDatumIterator{}
-	for _, input := range cross {
-		datumIterator, err := NewDatumIterator(pachClient, input)
+	defer result.Reset()
+	for _, iterator := range cross {
+		datumIterator, err := NewDatumIterator(pachClient, iterator)
 		if err != nil {
 			return nil, err
 		}
-		// start the iterator, and make sure it isn't empty
-		if !datumIterator.Next() {
-			return &crossDatumIterator{}, nil
-		}
-		result.inputs = append(result.inputs, datumIterator)
+		result.iterators = append(result.iterators, datumIterator)
 	}
 	return result, nil
 }
 
 func (d *crossDatumIterator) Reset() {
-	for _, input := range d.inputs {
-		input.Reset()
-		input.Next()
+	inhabited := true
```

Comment:
I think we are going to run into a problem when using this iterator in a composition. I believe `Reset` would get called twice (once when we create the cross iterator, and once when we create the iterator that has the cross within it) which would result with started being set to false after the second call when the cross should produce no datums. The first call will set the iterators to nil, and the second call would never set inhabited to false (therefore making started false). If think something like this should work:
```
for _, iterator := range d.iterators {
	iterator.Reset()
	if !iterator.Next() {
		d.iterators = nil
	}
}
d.started = len(d.iterators) == 0
```

Reply:
Also, the updated test looks good. I would just add a test for this particular case (cross that produces zero datums embedded in a union or cross).